### PR TITLE
#650 - Auto Complete field doesn't show suggestion for concept linking

### PR DESCRIPTION
**Pull request by reckart**
- Introduce constant for query lower limit
- Fix maxResults = 0 in the DB on application start
- Apply maxResults caps during project import
- Set default value if maxResults = 0 on import
- Use a NumberTextField for the maxResults in the project settings instead of a RequiredTextField

### CODE REVIEW

**Comment by naveen2507** [comment]
When I go for edit the KB, it throws internal error 

`
2018-11-21 09:28:59 ERROR [admin] DefaultExceptionMapper - Unexpected error occurred
org.apache.wicket.markup.MarkupException: Component [maxResults] (path = [14:tabContainer:tabPanel:panel:details:form:content:common-content:maxResults]) must be applied to a tag with [type] attribute matching any of [number], not [text]
	at org.apache.wicket.markup.MarkupStream.throwMarkupException(MarkupStream.java:534) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.checkComponentTagAttribute(Component.java:3516) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.markup.html.form.TextField.onComponentTag(TextField.java:98) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.markup.html.form.NumberTextField.onComponentTag(NumberTextField.java:235) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.internalRenderComponent(Component.java:2420) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.onRender(MarkupContainer.java:1592) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.internalRender(Component.java:2287) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.render(Component.java:2218) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.renderNext(MarkupContainer.java:1453) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.renderAll(MarkupContainer.java:1653) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.renderComponentTagBody(MarkupContainer.java:1628) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.onComponentTagBody(MarkupContainer.java:1586) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.markup.html.panel.DefaultMarkupSourcingStrategy.onComponentTagBody(DefaultMarkupSourcingStrategy.java:70) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.internalRenderComponent(Component.java:2479) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.onRender(MarkupContainer.java:1592) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.internalRender(Component.java:2287) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.render(Component.java:2218) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.renderNext(MarkupContainer.java:1453) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.renderAll(MarkupContainer.java:1653) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.renderComponentTagBody(MarkupContainer.java:1628) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.onComponentTagBody(MarkupContainer.java:1586) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.markup.html.panel.FragmentMarkupSourcingStrategy.onComponentTagBody(FragmentMarkupSourcingStrategy.java:89) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.internalRenderComponent(Component.java:2479) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.onRender(MarkupContainer.java:1592) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.internalRender(Component.java:2287) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.render(Component.java:2218) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.renderNext(MarkupContainer.java:1453) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.renderAll(MarkupContainer.java:1653) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.renderComponentTagBody(MarkupContainer.java:1628) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.onComponentTagBody(MarkupContainer.java:1586) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.markup.html.form.Form.onComponentTagBody(Form.java:1667) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.markup.html.panel.DefaultMarkupSourcingStrategy.onComponentTagBody(DefaultMarkupSourcingStrategy.java:70) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.internalRenderComponent(Component.java:2479) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.onRender(MarkupContainer.java:1592) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.internalRender(Component.java:2287) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.render(Component.java:2218) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.renderNext(MarkupContainer.java:1453) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.renderAll(MarkupContainer.java:1653) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.renderComponentTagBody(MarkupContainer.java:1628) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.renderAssociatedMarkup(MarkupContainer.java:797) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.markup.html.panel.AssociatedMarkupSourcingStrategy.renderAssociatedMarkup(AssociatedMarkupSourcingStrategy.java:77) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.markup.html.panel.PanelMarkupSourcingStrategy.onComponentTagBody(PanelMarkupSourcingStrategy.java:112) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.internalRenderComponent(Component.java:2479) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.MarkupContainer.onRender(MarkupContainer.java:1592) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.internalRender(Component.java:2287) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.render(Component.java:2218) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.Component.renderPart(Component.java:2195) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.page.XmlPartialPageUpdate.writeComponent(XmlPartialPageUpdate.java:75) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.page.PartialPageUpdate.writeComponents(PartialPageUpdate.java:263) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.page.PartialPageUpdate.writeTo(PartialPageUpdate.java:162) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.ajax.AjaxRequestHandler.respond(AjaxRequestHandler.java:384) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.request.cycle.RequestCycle$HandlerExecutor.respond(RequestCycle.java:912) ~[wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.request.RequestHandlerExecutor.execute(RequestHandlerExecutor.java:65) ~[wicket-request-8.1.0.jar:8.1.0]
	at org.apache.wicket.request.cycle.RequestCycle.execute(RequestCycle.java:283) [wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.request.cycle.RequestCycle.processRequest(RequestCycle.java:253) [wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.request.cycle.RequestCycle.processRequestAndDetach(RequestCycle.java:221) [wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.protocol.http.WicketFilter.processRequestCycle(WicketFilter.java:262) [wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.protocol.http.WicketFilter.processRequest(WicketFilter.java:204) [wicket-core-8.1.0.jar:8.1.0]
	at org.apache.wicket.protocol.http.WicketFilter.doFilter(WicketFilter.java:286) [wicket-core-8.1.0.jar:8.1.0]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [catalina.jar:8.5.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [catalina.jar:8.5.30]
	at org.springframework.orm.jpa.support.OpenEntityManagerInViewFilter.doFilterInternal(OpenEntityManagerInViewFilter.java:185) [spring-orm-5.1.1.RELEASE.jar:5.1.1.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) [spring-web-5.1.1.RELEASE.jar:5.1.1.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [catalina.jar:8.5.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [catalina.jar:8.5.30]
	at de.tudarmstadt.ukp.clarin.webanno.support.logging.LoggingFilter.doFilter(LoggingFilter.java:52) [webanno-support-3.5.0-beta-9.jar:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [catalina.jar:8.5.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [catalina.jar:8.5.30]
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:82) [spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [catalina.jar:8.5.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [catalina.jar:8.5.30]
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200) [spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [catalina.jar:8.5.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [catalina.jar:8.5.30]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:116) [spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91) [spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [catalina.jar:8.5.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [catalina.jar:8.5.30]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:116) [spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91) [spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [catalina.jar:8.5.30]
.....
`

**Comment by reckart** [comment]
@naveen2507 Sry - looking into it.

**Comment by reckart** [comment]
@naveen2507 should be good now :)

**Comment by naveen2507** [comment]
@reckart now we have the issue with document import here. I checked blinky-experimental but the issue doesn't exist in there. 

Version : 4083f7f966e34500b9a91b0f00b89403f928e55c

Steps to reproduce: 
1. Create a new project
2. Import document. 
3. In Annotation Window, the document doesn't appear. 

Ideally it should list the documents.

**Comment by reckart** [comment]
Any error in the console/log?

**Comment by naveen2507** [comment]
> Any error in the console/log?

Nopes

**Comment by naveen2507** [comment]
After rebuilding the index:
`
2018-11-21 14:54:53 INFO [admin] KnowledgeBaseDetailsPanel - Starting rebuilding full-text index of KnowledgeBase [id=pid-15-kbid-] ... this may take a while ...`
`2018-11-21 14:54:53 INFO [admin] KnowledgeBaseDetailsPanel - Completed rebuilding full-text index of KnowledgeBase [id=pid-15-kbid-]`

I don't get auto complete suggestions : 
2018-11-21 15:06:22 ERROR [admin] ConceptLinkingService - Mention not found in sentence!

image

**Comment by naveen2507** [comment]
@reckart does this issue depends on #560. As I am testing with Wine ontology which doesn't have explicit property for labels

**Comment by naveen2507** [comment]
The issue for document import appeared for me , but when I tried to reproduce it. It works.

**Comment by reckart** [comment]
> I don't get auto complete suggestions : 2018-11-21 15:06:22 ERROR [admin] ConceptLinkingService - Mention not found in sentence!

Interestingly, I do not get the error message - but I also get no auto-completed. That is: I get one when I type "s", namely "subClassOf" - but that doesn't make any sense at all...

**Comment by naveen2507** [comment]
yeah for now it make sense. But I would suggest not to close the issue #650 for now.
